src/extract/pathbuf.rs

```diff
@@ -1,116 +1,101 @@
-                let path = ctx.request().uri().path();
-                let value = ctx.captures().get(*idx, path).into();
-                ExtractPathBuf(value)
-            }
-            _ => unimplemented!("A PathBuf can only be extracted from a path capture for now"),
-        }
-    }
+#[cfg(not(any(target_os = "windows", target_arch = "wasm32")))]
+fn osstr_from_bytes(bytes: &[u8]) -> Result<&OsStr, Error> {
+    use std::os::unix::ffi::OsStrExt;
+    Ok(OsStr::from_bytes(bytes))
 }
 
-/// Extracts a `PathBuf` value from an HTTP request.
-#[derive(Debug)]
-pub struct ExtractPathBuf(String);
+#[cfg(any(target_os = "windows", target_arch = "wasm32"))]
+fn osstr_from_bytes(bytes: &[u8]) -> Result<&OsStr, Error> {
+    use std::str;
+    str::from_utf8(bytes)
+        .map_err(|e| Error::invalid_argument(&e))
+        .map(|s| OsStr::new(s))
+}
 
-impl ExtractPathBuf {
-    // https://www.owasp.org/index.php/Path_Traversal
-    fn check_for_path_traversal(&self) -> Result<(), Error> {
-        use self::path::Component::*;
+// https://www.owasp.org/index.php/Path_Traversal
+fn check_for_path_traversal(path: &Path) -> Result<(), Error> {
+    use self::path::Component::*;
 
-        let path_traversal_error = || Error::invalid_argument(&"Path traversal detected");
+    let path_traversal_error = || Error::invalid_argument(&"Path traversal detected");
 
-        let mut depth: u32 = 0;
-        for c in Path::new(&self.0).components() {
-            match c {
-                Prefix(_) | RootDir => {
-                    // Escaping to the root is immediately a failure
-                    return Err(path_traversal_error());
-                }
-                CurDir => {
-                    // no-op
-                }
-                ParentDir => {
-                    depth = match depth.checked_sub(1) {
-                        Some(v) => v,
-                        None => return Err(path_traversal_error()),
-                    }
-                }
-                Normal(_) => {
-                    depth += 1;
+    let mut depth = 0u32;
+    for c in path.components() {
+        match c {
+            Prefix(_) | RootDir => {
+                // Escaping to the root is immediately a failure
+                Err(path_traversal_error())?
+            }
+            CurDir => {
+                // no-op
+            }
+            ParentDir => {
+                depth = match depth.checked_sub(1) {
+                    Some(v) => v,
+                    None => Err(path_traversal_error())?,
                 }
             }
+            Normal(_) => {
+                depth += 1;
+            }
         }
-
-        Ok(())
     }
+
+    Ok(())
 }
 
-impl ExtractFuture for ExtractPathBuf {
-    type Item = PathBuf;
+fn decode(s: &str) -> Result<PathBuf, Error> {
+    let percent_decoded = Cow::from(percent_encoding::percent_decode(s.as_bytes()));
+    let path = PathBuf::from(osstr_from_bytes(percent_decoded.as_ref())?);
```

Comment:
Perhaps this code would be better off as an implementation of `Extract` for `OsString`?

Reply:
Yeah, an `OsString` impl might be a good idea, I'll try that later.

Actually, it might be even better to have a different `Extract` trait for types than can be extracted immediately (see https://github.com/carllerche/tower-web/issues/105#issuecomment-419597212), but that's a different story.

Reply:
It looks a bit awkward, but done.